### PR TITLE
sarif validation rule 2010 - provide code snippets

### PR DESCRIPTION
**Pull request by harleenkohli**

### CODE REVIEW

File: src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/ExpectedOutputs/SARIF2010.ProvideCodeSnippets_Invalid.sarif [hunk 1]
```diff
@@ -0,0 +1,45 @@
+{
+  "$schema": "https://schemastore.azurewebsites.net/schemas/json/sarif-2.1.0-rtm.5.json",
+  "version": "2.1.0",
```
Comment (harleenkohli) [inline src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/ExpectedOutputs/SARIF2010.ProvideCodeSnippets_Invalid.sarif:3]: oops, wrong file, pushing the correct one! hold on!

File: src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF2010.ProvideCodeSnippets_Invalid.sarif [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                  "uri": "src/test.c"
+                },
+                "region": {
+                  "startLine": 3
+                },
+                "contextRegion": {
+                  "startLine": 2,
+                  "endLine": 4
+                }
+              }
```
Comment (lgolding) [inline src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF2010.ProvideCodeSnippets_Invalid.sarif:37]: >} [](start = 14, length = 1)

Add test cases where:

- There is only a region, and it doesn't have a snippet.
- There are both a region and a contextRegion, region has a snippet but contextRegion does not.
- And vice versa.

File: src/Sarif.Multitool/Rules/SARIF2010.ProvideCodeSnippets.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                }
+            }
+        }
+
+        private void AnalyzeResultLocation(Location location, string locationPointer)
+        {
+            Region region = location.PhysicalLocation?.Region;
+            Region contextRegion = location.PhysicalLocation?.ContextRegion;
+
+            if (region != null && contextRegion != null)
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SARIF2010.ProvideCodeSnippets.cs:46]: >&& [](start = 31, length = 2)

Bug: should be logical OR.

File: src/Sarif.Multitool/Rules/SARIF2010.ProvideCodeSnippets.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                location.PhysicalLocation?.ContextRegion,
+                locationPointer
+                    .AtProperty(SarifPropertyName.PhysicalLocation)
+                    .AtProperty(SarifPropertyName.ContextRegion));
+        }
+
+        private void AnalyzeRegion(Region region, string regionPointer)
+        {
+
+            if (region != null && region.Snippet == null)
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SARIF2010.ProvideCodeSnippets.cs:59]: >if [](start = 12, length = 2)

Remove extra blank line above.

File: src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF2010.ProvideCodeSnippets_Valid.sarif [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                  "startLine": 3,
+                  "snippet": {"text": "region code snippet"}
+                },
+                "contextRegion": {
+                  "startLine": 2,
+                  "endLine": 4,
+                  "snippet": { "text": "context region code snippet. More context here." }
+                }
+              }
+            }
```
Comment (lgolding) [inline src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF2010.ProvideCodeSnippets_Valid.sarif:40]: >} [](start = 12, length = 1)

Add a valid test case where it's just a region with a snippet, no contextRegion.

Comment (lgolding) [reply src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF2010.ProvideCodeSnippets_Valid.sarif:40]: I take it back. That will cause a `result` to appear in your 'Valid' output file.

---
In reply to: [448028963](https://github.com/microsoft/sarif-sdk/pull/1953#discussion_r448028963) [](ancestors = 448028963)

Comment (harleenkohli) [reply src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF2010.ProvideCodeSnippets_Valid.sarif:40]: larry i added it and trhen decided to remove it - reason: then we get violation for rule#2011. it was basically  a copy of 2011_invalid. i can add it back - it just felt unnecessary. I can add it back.

---
In reply to: [448028963](https://github.com/microsoft/sarif-sdk/pull/1953#discussion_r448028963) [](ancestors = 448028963)

Comment (harleenkohli) [reply src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF2010.ProvideCodeSnippets_Valid.sarif:40]: aah thank u :) u read my mind!

---
In reply to: [448029654](https://github.com/microsoft/sarif-sdk/pull/1953#discussion_r448029654) [](ancestors = 448029654,448028963)

**Pull request merged**